draft-ietf-ppm-dap.md

```diff
@@ -2587,6 +2595,62 @@ the task and all data pertaining to this task after `task_expiration`.
 Implementors SHOULD provide for some leeway so the Collector can collect the
 batch after some delay.
 
+### Distributed-systems and Synchronization Concerns
+
+Various parts of a DAP implementation will need to synchronize in order to
+ensure correctness during concurrent operation. This section describes the
+relevant concerns and makes suggestions as to potential implementation
+tradeoffs.
+
+* The upload interaction requires the Leader to ignore uploaded reports with a
+  duplicated ID, including concurrently-uploaded reports. This might be
+  implemented by synchronization or via an eventually-consistent process. If the
+  Leader wishes to alert the Client with a `reportRejected` error,
+  synchronization will be necessary to ensure all but one concurrent request
+  receive the error.
+
+* The Leader is responsible for generating aggregation jobs, and will generally
+  want to place each report in exactly one aggregation job. This may require
```

Comment:
This suggests to me there are situations where a Leader would stick a report in two jobs. When would this happen?

Reply:
* If a report is rejected due to `report_too_early`, the Leader can place it in a later aggregation job. (this is touched on by the implementation note added in a late commit to this PR)
* In general, it is _acceptable_ to place a report in multiple aggregation jobs, as long as the Leader also performs report-replay checking later. This results in a loss of throughput but no correctness/privacy/robustness issues; it may be the right tradeoff to allow aggregation job creation to generate aggregation jobs with duplicated reports in some corner cases. (this is touched on by the remainder of this paragraph)

Reply:
> * If a report is rejected due to `report_too_early`, the Leader can place it in a later aggregation job. (this is touched on by the implementation note added in a late commit to this PR)

Let's make a note of this here.


>     * In general, it is _acceptable_ to place a report in multiple aggregation jobs, as long as the Leader also performs report-replay checking later. This results in a loss of throughput but no correctness/privacy/robustness issues; it may be the right tradeoff to allow aggregation job creation to generate aggregation jobs with duplicated reports in some corner cases. (this is touched on by the remainder of this paragraph)

Does this happen in your experience, or is this just something you imagine could happen? This sounds like an extreme edge case to me, and I worry that trying to make room for it here might mislead readers.

Reply:
> Let's make a note of this here.

👍🏻 

> Does this happen in your experience, or is this just something you imagine could happen? This sounds like an extreme edge case to me, and I worry that trying to make room for it here might mislead readers.

Janus doesn't currently take advantage of this; the Leader implementation will generate aggregation jobs which partition the space of reports. I do think it's useful to note that eventual consistency is acceptable here; for example, if aggregation jobs are generated by pulling reports from a distributed queue of pending reports, any time the distributed queue accidentally delivered a message multiple times we would see the same report showing up in multiple aggregation jobs -- and that's OK.